File: legacy/android/kotlin/StorylyTwilioIntegrationSample/app/src/main/java/com/appsamurai/storylytwiliosample/SegmentEventMapper.kt
package com.appsamurai.storylytwiliosample

import com.appsamurai.storyly.Story
import com.appsamurai.storyly.StoryButtonComponent
import com.appsamurai.storyly.StoryCommentComponent
import com.appsamurai.storyly.StoryComponent
import com.appsamurai.storyly.StoryEmojiComponent
import com.appsamurai.storyly.StoryGroup
import com.appsamurai.storyly.StoryGroupBadgeStyle
import com.appsamurai.storyly.StoryGroupStyle
import com.appsamurai.storyly.StoryImageQuizComponent
import com.appsamurai.storyly.StoryPollComponent
import com.appsamurai.storyly.StoryProductCardComponent
import com.appsamurai.storyly.StoryProductCatalogComponent
import com.appsamurai.storyly.StoryProductTagComponent
import com.appsamurai.storyly.StoryPromoCodeComponent
import com.appsamurai.storyly.StoryQuizComponent
import com.appsamurai.storyly.StoryRatingComponent
import com.appsamurai.storyly.StorySwipeComponent
import com.appsamurai.storyly.analytics.StorylyEvent
import com.appsamurai.storyly.data.managers.product.STRCart
import com.appsamurai.storyly.data.managers.product.STRCartItem
import com.appsamurai.storyly.data.managers.product.STRProductItem
import com.appsamurai.storyly.data.managers.product.STRProductVariant

/**
 * Builds a Segment `track` properties map from the objects that Storyly hands
 * back in the [com.appsamurai.storyly.StorylyListener] /
 * [com.appsamurai.storyly.StorylyProductListener] callbacks.
 *
 * Segment's Track spec: https://www.twilio.com/docs/segment/connections/spec/track
 * The story group, story, and interactive component each become a nested
 * sub-map, and repeated objects (products, variants, colors) become nested
 * lists. The result is a plain [Map] that [SegmentManager] converts to a
 * JsonObject before sending. Optional values are only written when present.
 *
 * Goal: full coverage of the public properties Storyly exposes.
 */
object SegmentEventMapper {

    // region Entry points

    /**
     * Assembles story & interactive context into a nested properties map.
     *
     * Shape:
     * ```
     * {
     *   "event": "StoryQuizAnswered",
     *   "event_raw_value": 20,
     *   "story_group": { ..., "style": { "badge": { ... } } },
     *   "story":       { ..., "action_products": [ { ..., "variants": [ ... ] } ] },
     *   "component":   { ..., "products": [ ... ] }
     * }
     * ```
     */
    fun properties(
        event: StorylyEvent?,
        storyGroup: StoryGroup?,
        story: Story?,
        storyComponent: StoryComponent?
    ): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>()
        event?.let { map.putAll(eventProperties(it)) }
        storyGroup?.let { map["story_group"] = storyGroupMap(it) }
        story?.let { map["story"] = storyMap(it) }
        storyComponent?.let { map["component"] = componentMap(it) }
        return map
    }

    /** Assembles a cart-related product event (from `storylyUpdateCartEvent`). */
    fun cartProperties(event: StorylyEvent, cart: STRCart?, change: STRCartItem?): Map<String, Any?> {
        val map = eventProperties(event).toMutableMap()
        cart?.let { map["cart"] = cartMap(it) }
        change?.let { map["change"] = cartItemMap(it) }
        return map
    }

    /** Assembles a wishlist-related product event (from `storylyUpdateWishlistEvent`). */
    fun wishlistProperties(event: StorylyEvent, item: STRProductItem?): Map<String, Any?> {
        val map = eventProperties(event).toMutableMap()
        item?.let { map["product"] = productMap(it) }
        return map
    }

    // endregion

    // region Story group

    private fun storyGroupMap(storyGroup: StoryGroup): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>(
            "id" to storyGroup.uniqueId,
            "title" to storyGroup.title,
            "index" to storyGroup.index,
            "seen" to storyGroup.seen,
            "pinned" to storyGroup.pinned,
            "nudge" to storyGroup.nudge,
            "type" to storyGroup.type.customName,
            "story_count" to storyGroup.stories.size
        )
        storyGroup.name?.let { map["name"] = it }
        storyGroup.iconUrl?.let { map["icon_url"] = it }
        storyGroup.iconVideoUrl?.let { map["icon_video_url"] = it }
        storyGroup.iconVideoThumbnailUrl?.let { map["icon_video_thumbnail_url"] = it }
        styleMap(storyGroup.style)?.let { map["style"] = it }
        return map
    }

    private fun styleMap(style: StoryGroupStyle?): Map<String, Any?>? {
        if (style == null) return null
        val map = mutableMapOf<String, Any?>()
        style.borderUnseenColors?.let { colors -> map["border_unseen_colors"] = colors.map { hexString(it) } }
        style.textUnseenColor?.let { map["text_unseen_color"] = hexString(it) }
        badgeMap(style.badge)?.let { map["badge"] = it }
        return map.ifEmpty { null }
    }

    private fun badgeMap(badge: StoryGroupBadgeStyle?): Map<String, Any?>? {
        if (badge == null) return null
        val map = mutableMapOf<String, Any?>()
        badge.text?.let { map["text"] = it }
        badge.textColor?.let { map["text_color"] = hexString(it) }
        badge.backgroundColor?.let { map["background_color"] = hexString(it) }
        badge.endTime?.let { map["end_time"] = it }
        badge.template?.let { map["template"] = it }
        return map.ifEmpty { null }
    }

    // endregion

    // region Story

    private fun storyMap(story: Story): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>(
            "id" to story.uniqueId,
            "title" to story.title,
            "index" to story.index,
            "seen" to story.seen
        )
        story.currentTime?.let { map["current_time"] = it }
        story.name?.let { map["name"] = it }
        story.actionUrl?.let { map["action_url"] = it }
        story.previewUrl?.let { map["preview_url"] = it }
        story.storyComponentList?.let { map["component_count"] = it.size }
        productMaps(story.actionProducts)?.let { map["action_products"] = it }
        return map
    }

    // endregion

    // region Interactive components

    /**
     * Builds the base component fields, then casts to each concrete type and
     * merges in its type-specific fields (including attached products).
     */
    private fun componentMap(component: StoryComponent): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>(
            "id" to component.id,
            "type" to component.type.name
        )
        component.customPayload?.let { map["custom_payload"] = it }

        when (component) {
            is StoryQuizComponent -> {
                map["title"] = component.title
                map["options"] = component.options
                map["selected_index"] = component.selectedOptionIndex
                component.rightAnswerIndex?.let { map["right_answer_index"] = it }
            }
            is StoryImageQuizComponent -> {
                component.title?.let { map["title"] = it }
                map["selected_index"] = component.selectedOptionIndex
                component.options?.let { map["options"] = it }
                component.rightAnswerIndex?.let { map["right_answer_index"] = it }
            }
            is StoryPollComponent -> {
                map["title"] = component.title
                map["options"] = component.options
                map["selected_index"] = component.selectedOptionIndex
            }
            is StoryEmojiComponent -> {
                map["emoji_codes"] = component.emojiCodes
                map["selected_index"] = component.selectedEmojiIndex
            }
            is StoryRatingComponent -> {
                map["emoji_code"] = component.emojiCode
                map["rating"] = component.rating
            }
            is StoryPromoCodeComponent -> {
                map["promo_code"] = component.text
            }
            is StoryCommentComponent -> {
                map["comment_text"] = component.text
            }
            is StorySwipeComponent -> {
                map["text"] = component.text
                component.actionUrl?.let { map["action_url"] = it }
                productMaps(component.products)?.let { map["products"] = it }
            }
            is StoryButtonComponent -> {
                map["text"] = component.text
                component.actionUrl?.let { map["action_url"] = it }
                productMaps(component.products)?.let { map["products"] = it }
            }
            is StoryProductCardComponent -> {
                component.text?.let { map["text"] = it }
                component.actionUrl?.let { map["action_url"] = it }
                productMaps(component.products)?.let { map["products"] = it }
            }
            is StoryProductTagComponent -> {
                component.actionUrl?.let { map["action_url"] = it }
                productMaps(component.products)?.let { map["products"] = it }
            }
            is StoryProductCatalogComponent -> {
                component.actionUrlList?.let { urls -> map["action_urls"] = urls.filterNotNull() }
                productMaps(component.products)?.let { map["products"] = it }
            }
        }
        return map
    }

    // endregion

    // region Products & cart

    private fun productMaps(products: List<STRProductItem>?): List<Map<String, Any?>>? {
        if (products.isNullOrEmpty()) return null
        return products.map { productMap(it) }
    }

    private fun productMap(product: STRProductItem): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>(
            "id" to product.productId,
            "group_id" to product.productGroupId,
            "title" to product.title,
            "url" to product.url,
            "price" to product.price,
            "currency" to product.currency,
            "wishlist" to (product.wishlist ?: false)
        )
        product.desc?.let { map["description"] = it }
        product.salesPrice?.let { map["sales_price"] = it }
        product.lowestPrice?.let { map["lowest_price"] = it }
        product.imageUrls?.let { map["image_urls"] = it }
        product.ctaText?.let { map["cta_text"] = it }
        variantMaps(product.variants)?.let { map["variants"] = it }
        return map
    }

    private fun variantMaps(variants: List<STRProductVariant>?): List<Map<String, Any?>>? {
        if (variants.isNullOrEmpty()) return null
        return variants.map { mapOf("name" to it.name, "value" to it.value, "key" to it.key) }
    }

    private fun cartMap(cart: STRCart): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>(
            "total_price" to cart.totalPrice,
            "currency" to cart.currency,
            "item_count" to cart.items.size,
            "items" to cart.items.map { cartItemMap(it) }
        )
        cart.oldTotalPrice?.let { map["old_total_price"] = it }
        return map
    }

    private fun cartItemMap(cartItem: STRCartItem): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>(
            "quantity" to cartItem.quantity,
            "product" to productMap(cartItem.item)
        )
        cartItem.totalPrice?.let { map["total_price"] = it }
        cartItem.oldTotalPrice?.let { map["old_total_price"] = it }
        return map
    }

    // endregion

    // region Helpers

    private fun eventProperties(event: StorylyEvent): Map<String, Any?> =
        mapOf("event" to event.name, "event_raw_value" to event.ordinal)

    /** Converts an Android ColorInt into an `#AARRGGBB` hex string. */
    private fun hexString(color: Int): String = String.format("#%08X", color)

    // endregion
}

File: legacy/ios/swift/StorylyTwilioIntegrationSample/StorylyTwilioIntegrationSample/SegmentEventMapper.swift
//
//  SegmentEventMapper.swift
//  StorylyTwilioIntegrationSample
//
//  Builds a Segment `track` properties dictionary from the objects that
//  Storyly hands back in the `storylyEvent` delegate callback.
//
//  Segment's Track spec: https://www.twilio.com/docs/segment/connections/spec/track
//  Properties are a free-form dictionary describing the action. Here we keep the
//  natural shape of the Storyly objects: the story group, story, and interactive
//  component each become a nested sub-dictionary, and repeated objects (products,
//  variants, colors) become nested arrays. Everything stays JSON-serializable.
//
//  Goal: full coverage of the public instance properties Storyly exposes.
//  Optional values are only written when present, so the map stays clean.
//

import Foundation
import UIKit
import Storyly

enum SegmentEventMapper {

    // MARK: - Entry point

    /// Assembles all available context from the `storylyEvent` callback into a
    /// nested properties dictionary.
    ///
    /// The `event` parameter is optional and defaults to `nil` so existing call
    /// sites that only pass the story context keep compiling. Pass it to include
    /// the event name/index in the properties as well.
    ///
    /// Shape:
    /// ```
    /// {
    ///   "event": "StoryQuizAnswered",
    ///   "event_raw_value": 20,
    ///   "story_group": { ..., "style": { "badge": { ... } } },
    ///   "story":       { ..., "action_products": [ { ..., "variants": [ ... ] } ] },
    ///   "component":   { ..., "products": [ ... ] }
    /// }
    /// ```
    static func properties(event: StorylyEvent? = nil,
                           storyGroup: StoryGroup?,
                           story: Story?,
                           storyComponent: StoryComponent?) -> [String: Any] {
        var props: [String: Any] = [:]

        if let event = event { props.merge(eventProperties(event)) { _, new in new } }
        if let storyGroup = storyGroup { props["story_group"] = storyGroupMap(storyGroup) }
        if let story = story { props["story"] = storyMap(story) }
        if let storyComponent = storyComponent { props["component"] = componentMap(storyComponent) }

        return props
    }

    /// Assembles a cart-related product event (from `storylyUpdateCartEvent`)
    /// into a nested properties dictionary.
    ///
    /// - Parameters:
    ///   - event: The cart event (e.g. StoryProductCartAdded, StoryProductAdded).
    ///   - cart: The current cart snapshot, if available.
    ///   - change: The single cart item being added/updated/removed, if available.
    static func cartProperties(event: StorylyEvent,
                               cart: STRCart?,
                               change: STRCartItem?) -> [String: Any] {
        var props = eventProperties(event)
        if let cart = cart { props["cart"] = cartMap(cart) }
        if let change = change { props["change"] = cartItemMap(change) }
        return props
    }

    /// Assembles a wishlist-related product event (from
    /// `storylyUpdateWishlistEvent`) into a nested properties dictionary.
    ///
    /// - Parameters:
    ///   - event: The wishlist event (e.g. StoryWishlistAdded, StoryWishlistRemoved).
    ///   - item: The product being added to / removed from the wishlist, if available.
    static func wishlistProperties(event: StorylyEvent,
                                   item: STRProductItem?) -> [String: Any] {
        var props = eventProperties(event)
        if let item = item { props["product"] = productMap(item) }
        return props
    }

    // MARK: - Story group

    private static func storyGroupMap(_ storyGroup: StoryGroup) -> [String: Any] {
        var map: [String: Any] = [
            "id": storyGroup.uniqueId,
            "title": storyGroup.title,
            "index": storyGroup.index,
            "seen": storyGroup.seen,
            "pinned": storyGroup.pinned,
            "nudge": storyGroup.nudge,
            "type": storyGroup.type.description,
            "story_count": storyGroup.stories.count
        ]

        if let name = storyGroup.name { map["name"] = name }
        if let url = storyGroup.iconUrl?.absoluteString { map["icon_url"] = url }
        if let url = storyGroup.iconVideoUrl?.absoluteString { map["icon_video_url"] = url }
        if let url = storyGroup.iconVideoThumbnailUrl?.absoluteString { map["icon_video_thumbnail_url"] = url }
        if let style = styleMap(storyGroup.style) { map["style"] = style }

        return map
    }

    private static func styleMap(_ style: StoryGroupStyle?) -> [String: Any]? {
        guard let style = style else { return nil }

        var map: [String: Any] = [:]
        if let colors = style.borderUnseenColors { map["border_unseen_colors"] = colors.map(hexString(from:)) }
        if let color = style.textUnseenColor { map["text_unseen_color"] = hexString(from: color) }
        if let badge = badgeMap(style.badge) { map["badge"] = badge }

        return map.isEmpty ? nil : map
    }

    private static func badgeMap(_ badge: StoryGroupBadgeStyle?) -> [String: Any]? {
        guard let badge = badge else { return nil }

        var map: [String: Any] = [:]
        if let text = badge.text { map["text"] = text }
        if let color = badge.textColor { map["text_color"] = hexString(from: color) }
        if let color = badge.backgroundColor { map["background_color"] = hexString(from: color) }
        if let endTime = badge.endTime { map["end_time"] = endTime.doubleValue }
        if let template = badge.template { map["template"] = template }

        return map.isEmpty ? nil : map
    }

    // MARK: - Story

    private static func storyMap(_ story: Story) -> [String: Any] {
        var map: [String: Any] = [
            "id": story.uniqueId,
            "title": story.title,
            "index": story.index,
            "seen": story.seen,
            "current_time": story.currentTime
        ]

        if let name = story.name { map["name"] = name }
        if let actionUrl = story.actionUrl { map["action_url"] = actionUrl }
        if let previewUrl = story.previewUrl?.absoluteString { map["preview_url"] = previewUrl }
        if let components = story.storyComponentList { map["component_count"] = components.count }
        if let products = productMaps(story.actionProducts) { map["action_products"] = products }

        return map
    }

    // MARK: - Interactive components

    /// Builds the base component fields, then casts to each concrete type and
    /// merges in its type-specific fields (including attached products).
    private static func componentMap(_ component: StoryComponent) -> [String: Any] {
        var map: [String: Any] = [
            "id": component.id,
            "type": component.type.stringValue
        ]
        if let payload = component.customPayload { map["custom_payload"] = payload }

        switch component {
        case let quiz as StoryQuizComponent:
            map["title"] = quiz.title
            map["options"] = quiz.options
            map["selected_index"] = quiz.selectedOptionIndex
            if let right = quiz.rightAnswerIndex { map["right_answer_index"] = right.intValue }

        case let quiz as StoryImageQuizComponent:
            map["title"] = quiz.title
            map["selected_index"] = quiz.selectedOptionIndex
            if let options = quiz.options { map["options"] = options }
            if let right = quiz.rightAnswerIndex { map["right_answer_index"] = right.intValue }

        case let poll as StoryPollComponent:
            map["title"] = poll.title
            map["options"] = poll.options
            map["selected_index"] = poll.selectedOptionIndex

        case let emoji as StoryEmojiComponent:
            map["emoji_codes"] = emoji.emojiCodes
            map["selected_index"] = emoji.selectedEmojiIndex

        case let rating as StoryRatingComponent:
            map["emoji_code"] = rating.emojiCode
            map["rating"] = rating.rating

        case let promo as StoryPromoCodeComponent:
            map["promo_code"] = promo.text

        case let comment as StoryCommentComponent:
            map["comment_text"] = comment.text

        case let swipe as StorySwipeComponent:
            map["text"] = swipe.text
            if let url = swipe.actionUrl { map["action_url"] = url }
            if let products = productMaps(swipe.products) { map["products"] = products }

        case let button as StoryButtonComponent:
            map["text"] = button.text
            if let url = button.actionUrl { map["action_url"] = url }
            if let products = productMaps(button.products) { map["products"] = products }

        case let productCard as StoryProductCardComponent:
            map["text"] = productCard.text
            if let url = productCard.actionUrl { map["action_url"] = url }
            if let products = productMaps(productCard.products) { map["products"] = products }

        case let productTag as StoryProductTagComponent:
            if let url = productTag.actionUrl { map["action_url"] = url }
            if let products = productMaps(productTag.products) { map["products"] = products }

        case let catalog as StoryProductCatalogComponent:
            if let urls = catalog.actionUrlList { map["action_urls"] = urls }
            if let products = productMaps(catalog.products) { map["products"] = products }

        default:
            break
        }

        return map
    }

    // MARK: - Products

    private static func productMaps(_ products: [STRProductItem]?) -> [[String: Any]]? {
        guard let products = products, !products.isEmpty else { return nil }
        return products.map(productMap(_:))
    }

    private static func productMap(_ product: STRProductItem) -> [String: Any] {
        var map: [String: Any] = [
            "id": product.productId,
            "group_id": product.productGroupId,
            "title": product.title,
            "url": product.url,
            "price": product.price,
            "currency": product.currency,
            "wishlist": product.wishlist
        ]

        if let desc = product.desc { map["description"] = desc }
        if let salesPrice = product.salesPrice { map["sales_price"] = salesPrice.floatValue }
        if let lowestPrice = product.lowestPrice { map["lowest_price"] = lowestPrice.floatValue }
        if let imageUrls = product.imageUrls { map["image_urls"] = imageUrls }
        if let ctaText = product.ctaText { map["cta_text"] = ctaText }
        if let variants = variantMaps(product.variants) { map["variants"] = variants }

        return map
    }

    private static func variantMaps(_ variants: [STRProductVariant]?) -> [[String: Any]]? {
        guard let variants = variants, !variants.isEmpty else { return nil }
        return variants.map { variant in
            ["name": variant.name, "value": variant.value, "key": variant.key]
        }
    }

    // MARK: - Cart

    private static func cartMap(_ cart: STRCart) -> [String: Any] {
        var map: [String: Any] = [
            "total_price": cart.totalPrice,
            "currency": cart.currency,
            "item_count": cart.items.count,
            "items": cart.items.map(cartItemMap(_:))
        ]
        if let oldTotalPrice = cart.oldTotalPrice { map["old_total_price"] = oldTotalPrice.floatValue }
        return map
    }

    private static func cartItemMap(_ cartItem: STRCartItem) -> [String: Any] {
        var map: [String: Any] = [
            "quantity": cartItem.quantity,
            "product": productMap(cartItem.item)
        ]
        if let totalPrice = cartItem.totalPrice { map["total_price"] = totalPrice.floatValue }
        if let oldTotalPrice = cartItem.oldTotalPrice { map["old_total_price"] = oldTotalPrice.floatValue }
        return map
    }

    // MARK: - Helpers

    private static func eventProperties(_ event: StorylyEvent) -> [String: Any] {
        ["event": event.stringValue, "event_raw_value": event.rawValue]
    }

    /// Converts a `UIColor` into a `#RRGGBBAA` hex string so colors survive
    /// JSON serialization.
    private static func hexString(from color: UIColor) -> String {
        var red: CGFloat = 0, green: CGFloat = 0, blue: CGFloat = 0, alpha: CGFloat = 0
        _ = color.getRed(&red, green: &green, blue: &blue, alpha: &alpha)
        let clamp: (CGFloat) -> Int = { Int((max(0, min(1, $0)) * 255).rounded()) }
        return String(format: "#%02X%02X%02X%02X", clamp(red), clamp(green), clamp(blue), clamp(alpha))
    }
}
